### PR TITLE
Edit user

### LINKED ISSUE #1446: User sub page with tabs and edit

### PR DESCRIPTION
**Pull request by na3d**
Closes #1446

### CODE REVIEW

**Comment by na3d** [comment]
screen shot 2017-03-24 at 10 04 40 am

File: src/users/components/UserForm.js [hunk 1]
```diff
@@ -10,12 +10,12 @@ import { ErrorSummary } from '~/errors';
...
 export class UserForm extends Component {
-  constructor() {
-    super();
+  constructor(props) {
+    super(props);
     this.formSubmit = this.formSubmit.bind(this);
     this.state = {
-      username: '',
-      email: '',
+      username: props.username || '',
```
Comment (mparke) [inline src/users/components/UserForm.js]: use `defaultProps` instead of `||` for consistency

**Comment by mparke** [comment]
minor alignment problem
screen shot 2017-03-24 at 10 28 56 am

**Comment by mparke** [comment]
console warning after edit save `Warning: flattenChildren(...): Encountered two children with the same key, `testusers1`. Child keys must be unique; when two children share a key, only the first child will be used.`

**Comment by na3d** [comment]
The flattenChildren problem is on the index page, because of an issue with state management.

File: test/users/user/layouts/EditUserPage.spec.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      username: 'theUser',
+      email: 'user@example.com',
+      password: 'password',
+    };
+    await page.instance().onSubmit(values);
+    expect(dispatch.calledTwice).to.equal(true);
+    const fn = dispatch.firstCall.args[0];
+    await expectRequest(
+      fn, `/account/users/${params.username}`,
+      undefined, undefined, {
```
Comment (abemassry) [inline test/users/user/layouts/EditUserPage.spec.js]: according to @eatonphil these aren't needed anymore

File: src/users/user/layouts/EditUserPage.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      this.setState({ errors, loading: false });
+    }
+  }
+
+  render() {
+    const { username } = this.props.params;
+    const { email } = this.props.users[username];
+    const errors = {};
+    return (
+      <div className="PrimaryPage container">
```
Comment (mparke) [inline src/users/user/layouts/EditUserPage.js]: remove `container` here to fix alignment

File: test/users/user/layouts/EditUserPage.spec.js (hunk 2 above)
Comment (na3d) [inline test/users/user/layouts/EditUserPage.spec.js]: I can remove one, it throws an error if I remove both.

File: src/users/user/layouts/EditUserPage.js (hunk 3 above)
Comment (mparke) [inline src/users/user/layouts/EditUserPage.js]: `container` creates a gutter, so the first child needs `row` I think, to fix the padding

File: test/users/user/layouts/EditUserPage.spec.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      username: 'theUser',
+      email: 'user@example.com',
+      password: 'password',
+    };
+    await page.instance().onSubmit(values);
+    expect(dispatch.calledTwice).to.equal(true);
+    const fn = dispatch.firstCall.args[0];
+    await expectRequest(
+      fn, `/account/users/${params.username}`,
+      undefined, {
```
Comment (eatonphil) [inline test/users/user/layouts/EditUserPage.spec.js]: With this undefined here, this test is definitely not doing what you expect.

Comment (eatonphil) [inline test/users/user/layouts/EditUserPage.spec.js]: expectRequest takes four arguments: the function to test, the url to test, the data the request is supposed to send, and the data the request should receive when it completes.

File: test/users/user/layouts/EditUserPage.spec.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    };
+    await page.instance().onSubmit(values);
+    expect(dispatch.calledTwice).to.equal(true);
+    const fn = dispatch.firstCall.args[0];
+    await expectRequest(
+      fn, `/account/users/${params.username}`, {
+        method: 'PUT',
+        body: { ...values },
+      }
+    );
```
Comment (eatonphil) [inline test/users/user/layouts/EditUserPage.spec.js]: +1

File: src/users/user/layouts/EditUserPage.js (hunk 3 above)
Comment (eatonphil) [inline src/users/user/layouts/EditUserPage.js]: +1

File: src/users/user/layouts/EditUserPage.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    username: PropTypes.string,
+  }),
+  username: PropTypes.string,
+  dispatch: PropTypes.func.isRequired,
+};
+
+function select(state) {
+  return {
+    users: state.api.users.users,
+    username: state.authentication.username,
```
Comment (eatonphil) [inline src/users/user/layouts/EditUserPage.js]: Why is this here?

Comment (na3d) [inline src/users/user/layouts/EditUserPage.js]: thank you!

File: test/users/user/layouts/EditUserPage.spec.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    dispatch.reset();
+    const values = {
+      username: 'theUser',
+      email: 'user@example.com',
+      password: 'password',
+    };
+    await page.instance().onSubmit(values);
+    expect(dispatch.calledTwice).to.equal(true);
```
Comment (eatonphil) [inline test/users/user/layouts/EditUserPage.spec.js]: Urgh! Don't compare things to true. It failed on this and we don't know how many dispatch calls there were. (There were three.) Please do `dispatch.callCount).to.equal(3);` instead.

Comment (na3d) [inline test/users/user/layouts/EditUserPage.spec.js]: Haha, habit, sorry.

File: src/users/user/layouts/EditUserPage.js [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      const errors = await reduceErrors(response);
+      this.setState({ errors, loading: false });
+    }
+  }
+
+  render() {
+    const { errors } = this.state;
+    const { username } = this.props.params;
+
+    // TODO: remove once primary key stops changing
```
Comment (mparke) [inline src/users/user/layouts/EditUserPage.js]: do we need an issue to revisit this?

Comment (eatonphil) [inline src/users/user/layouts/EditUserPage.js]: Hmm.. Does this break if the username doesn't change?

Comment (eatonphil) [inline src/users/user/layouts/EditUserPage.js]: https://github.com/linode/manager/issues/1465
